esti/copy_test.go

```diff
@@ -123,6 +127,79 @@ func TestCopyObject(t *testing.T) {
 		require.NoError(t, err)
 		require.Equal(t, http.StatusNotFound, statResp.StatusCode())
 	})
+
+	t.Run("read-only repository", func(t *testing.T) {
+		name := strings.ToLower(t.Name())
+		storageNamespace := generateUniqueStorageNamespace(name)
+		repoName := makeRepositoryName(name)
+		resp, err := client.CreateRepositoryWithResponse(ctx, &apigen.CreateRepositoryParams{}, apigen.CreateRepositoryJSONRequestBody{
+			DefaultBranch:    apiutil.Ptr(mainBranch),
+			Name:             repoName,
+			StorageNamespace: storageNamespace,
+			ReadOnly:         swag.Bool(true),
+		})
+		require.NoErrorf(t, err, "failed to create repository '%s', storage '%s'", name, storageNamespace)
+		require.NoErrorf(t, verifyResponse(resp.HTTPResponse, resp.Body),
+			"create repository '%s', storage '%s'", name, storageNamespace)
+		defer tearDownTest(repoName)
+
+		importPath := getImportPath(t)
+
+		const ingestionBranch = "test-copy"
+
+		_ = testImportNew(t, ctx, repoName, ingestionBranch,
+			[]apigen.ImportLocation{{Path: importPath, Type: "common_prefix"}},
+			map[string]string{"created_by": "import"},
+			true,
+		)
+
+		res, err := client.StatObjectWithResponse(ctx, repoName, ingestionBranch, &apigen.StatObjectParams{
+			Path: largeObject,
+		})
+		require.NoError(t, err)
+		require.Equal(t, http.StatusOK, res.StatusCode())
+
+		objStat := res.JSON200
+		destPath := "foo"
+		srcBranch := ingestionBranch
+		copyResp, err := client.CopyObjectWithResponse(ctx, repoName, "main", &apigen.CopyObjectParams{
+			DestPath: destPath,
+		}, apigen.CopyObjectJSONRequestBody{
+			SrcPath: largeObject,
+			SrcRef:  &srcBranch,
+		})
+		if copyResp.JSON403 == nil {
+			t.Fatalf("expected 403 forbidden error for CopyObject on read-only repository, got %d status code instead", copyResp.StatusCode())
```

Comment:
Here and elsewhere, use `require`.  There has to be some method to check for `nil` there.  I know, I don't like `require` anyway, but having _two_ types of test in the same file is even worse.